docs/adr/0055-communication-bounded-contexts.md

```diff
@@ -0,0 +1,61 @@
+
+Cependant, il reste la problématique des sujets qui sont à cheval entre deux équipes sur lesquels la frontière est floue.
+Ce code partagé est actuellement placé dans shared quand il a été identifié ou alors un bounded context importe directement le code d'autre bounded context.
+
+Il y a plusieurs conséquences négatives à garder ce code dans un état partagé : 
+- nécessite de la communication et du partage de connaissance
+- allonge le lead time 
+- perte d'engagement sur le sujet
+- perte d'autonomie des équipes
+
+Il est donc important de clarifier la responsabilité des équipes sur ces sujets.
+Une attribution claire facilitera les interactions entre les équipes et les rendra explicite dans la base de code.
+
+**Précision sur le terme "synchrone" du titre :**
+
+Il existe deux types d’interaction entre deux bounded contexts ou sub domains : 
+- J’ai besoin qu’un autre bounded context fasse une action pour moi ou me fournisse une information (ex : créer une campagne, lister les profils cibles d’une organisation, …)
+- J’ai besoin de réagir à un évènement qui se passe dans un autre bounded context (ex : un utilisateur a obtenu le niveau 1 dans une compétence, un utilisateur a remis à 0 une compétence, …)
+
+Dans cet ADR nous nous concentrons sur le premier type d’interaction.
+
+### Solution : Utiliser des APIs internes
+
+Il s'agit de définir une frontière entre les deux bounded contexts qui interagisse ensemble.
+En règle générale, un bounded context sera en position de fournir un service à un autre bounded context.
+L'API va venir expliciter ce service en définissant un contrat d'interface.
+
+Dans le cas où ces bounded contexts appartiennent à deux équipes différentes ce contrat d'interface va apporter plusieurs avantages : 
+- rendre les équipes autonomes sur le périmètre (tant que le contrat est respecté)
+- réduire la charge mentale des développeurs (réduit la somme de connaissances nécessaires pour réaliser les fonctionnalités)
+- clarifie l'attribution des sujets aux équipes (moins de discussions sur la prise en charge des sujets)
+
+Pour plus de détail sur l'implémentation de cette solution se référer à la [documentation](https://1024pix.atlassian.net/wiki/spaces/EDTDT/pages/3929735180/Comment+int+ragir+entre+deux+bounded+contexts+ou+des+sous+domain) dans Confluence.
+
+## Décision
+
+Les équipes ExpEval, Certif et Prescription ont déjà expérimenté la mise en place d'API internes.
+Lors de la phase d'expérimentation, il y a plusieurs retours négatifs qui ont été jugés acceptables (balance avantages/inconvénients) :
+- Boilerplate plus important dans la couche infrastructure avec l'ajout d'injection de dépendance dans les repositories
+- Sentiment de duplication de code tant que l'API n'est pas utilisée plusieurs fois
+- Utilisation du nommage Repository (côté consommateur) pour du code qui ne fait pas appel à la DB est source de confusion
```

Comment:
On n'a pas moyen d'utiliser un terme + adéquat ?

Reply:
La question revient assez souvent comme dans cet autre commentaire : https://github.com/1024pix/pix/pull/8511#discussion_r1557190975

Pour moi il faut qu'on clarifie ce qu'est le pattern Repository car il me semble adapté à la situation. Chez Pix, on a l'habitude de l'utiliser pour des considérations de base de données. Mais c'est une vision réductrice du pattern qui lui fait perdre tout son intérêt.

Le pattern Repository vient du DDD. Il permet de s'abstraire de la manière d'accéder aux données (DB, File System, HTTP, ... tout ce que vous pouvez imaginer). Il permet aussi de centraliser la génération d'objet du domain.

> A repository commonly refers to a storage location, usually considered a place of safety or preservation of the items stored in it. When you store something in a repository and later return to retrieve it, you expect that it will be in the same stage as it was in when you put it there. At some point, you may choose to remove the stored item from the repository.
>
> Implementing Domain-Driven Design - Chapter 12 - Vaughn Vernon

On voit bien qu'il n'y est pas question de base de données. Plus loin dans le chapitre, il parle des différentes implémentation possibles en fonction de où la donnée est stockée.

L'idée principale d'utiliser ce pattern dans le cadre de la Clean Architecture est que la couche domain ne soit pas au courant des considérations liés à la base de donnée ("The Database Is a Detail").

> To achieve this decoupling, we separate the UI of the add-order use case from the UI of the delete-order use case. We do the same with the business rules, and with the database. We keep the use cases separate down the vertical height if the system.
>
> You can see the pattern here. If you decouple elements of the system that change for different reasons, then you can continue to add new use cases without intefering with old ones.
> 
> Clean Architecture - Chapter 16 - Robert C. Martin

On voit ici que l'idée est que le code doit être modifié pour les bonnes raisons. On ne fait pas de SQL dans les usecases parce que si le schéma de base de données change on ne veut pas que ça impacte notre couche métier (puisqu'il n'a pas changé).

Si on en revient à notre situation vis à vis des APIs internes. C'est un détail d'où la donnée provient, c'est donc une considération de la couche infrastructure et qui plus est des Repositories qui vont pouvoir faire leur travail d'aller chercher la donnée au bon endroit et de se charger de l'aggréger pour produire un ou des objets du domain que le usecase appelant va pouvoir manipuler sans avoir conscience de leur provenance.

Reply:
Merci beaucoup ! C'est un peu + clair 👌 

Est-ce qu'on a un schema de l'architecture ? Ça pourrait aider à éclaircir notre façon d'utiliser les repositories ?

Reply:
Je peux faire ça effectivement, merci pour la suggestion 🙏

Reply:
Nous avons fait une revue ensemble cette après-midi avec @frinyvonnick , ce commentaire va répondre à mon commentaire précèdent pour tracer la discussion et la partager ici à tous.

Nous avons revue les ouvrages de référence sur les définitions
  * Interface Adapters : vocabulaire architecture hexagonal et non clean architecture
  * Gateway : est en fait "Database Gateways", vocabulaire très spécialisée
     * > « Between the use case interactors and the database are the database gateways. These gateways are polymorphic interfaces that contain methods for every create, read, update, or delete operation that can be performed by the application on the database. »
   * Repository  : terme générique non lié à une source de donnée particulière
     * > A repository commonly refers to a storage location, usually considered a place of safety or preservation of the items stored in it. When you store something in a repository and later return to retrieve it, you expect that it will be in the same stage as it was in when you put it there. At some point, you may choose to remove the stored item from the repository.
     * A la relecture de mon commentaire précèdent, on comprend que les termes "gateway" et "repository" en terme de nommage, représente globalement la même abstraction

---

Sur la base de ces définitions, reprenons point par point le commentaire ci-dessus
> a codebase actuelle est quand même + basée sur du model adapter + couche driver
elle contient au même niveau d'autres choses que des repository et qui n'ont pas le nom reposiotry
exemple : api/src/prescription/learner-management/infrastructure/storage/import-storage.js
Dans les autres cas on parle bien de "providers" et pas de repositories : OIDCProvider, MailProvider, S3ObjectStorageProvider, etc.

Conclusion : il y aura certainement un effort à faire suite à cet ADR pour renommer certains fichiers / classes dans la codebase actuelle, et en faveur du terme "repository". La couche métier actuelle, par les nommages qui existent aujourd'hui, ont conscience du type de source de données derrière, exemple: le "import-storage" devrait être "import-repository" pour respecter la définition précèdente

---

> que j'ai pas l'impression que le pole enginneering soit au courant que l'on pratique le repository pattern au sens DDD (vu que l'on parlait pas de DDD avant, et que par contre on a présenté mainte fois la clean archi)
>  
> que je trouve que c'est quand même un truc à part de cette notion d'API interne, certes compatibles, mais qui ajoute de la confusion (la remarque de mireille ici a été remonté plusieurs fois)
>
>que je suis pas sur que l'on implémente bien ce design pattern actuellement chez Pix

Conclusion : il serait intéressant de refaire une MEJ pour représenter le pattern Repository à l'ensemble de Pix, et éliminer les incertitudes qui pourraient exister sur la compréhension du concept de Repository pattern (définition DDD). Et en supplément : d'ajouter cela comme point à aborder lors des bootcamp pour l'onboarding des prochains devs.

---
> et que je trouve que pour l'adoption de la notion d'api interne (qui est le vrai sujet de cet ADR) je reste persuadé que le nom repository vient rendre la chose confusante (la preuve avec ce commentaire), et que le principe d'API interne vivrait très bien en ayant un nommage "plus simple".

Conclusion : ce point là n'existe plus si on fait l'effort de formation proposé juste avant. Ce qui ressort du point comme étant encore la meilleure solution.

---

Note personnelle : je préfère rester sur le nommage expliqué et défendu par les différents ADR et présenté ici dans cet ADR sur ce fil de commentaire !

Reply:
Je viens de me souvenir d'un exemple très concret d'implémentation changeant de Repositories en fonction de la source de données. Les Repositories (skill, competences, area, ...) liés au référentiel sont passés par plusieurs étapes sans avoir d'impact sur la couche domain.

Au début de Pix, ces repositories tappaient directement sur Airtable. On s'est retrouvé confronté à des soucis rate limiting. On a donc dû implémenter une couche de cache avec Redis et grâce à cette couche de repositories ça a été transparent. Et plus récemment, on a décidé d'avoir une API qui sert le référentiel qui s'appelle LCMS et encore une fois grâce aux repositories il n'y a toujours pas eu d'impact sur le domain. 

En conclusion, la source de données à changer 3 fois : 
- API js Airtable
- Redis
- API Node

Et aucune de ces sources de données n'est une base de données SQL 😄